src/app/WriteClient.cpp

```diff
@@ -244,19 +244,16 @@ CHIP_ERROR WriteClient::SendWriteRequest(SessionHandle session, System::Clock::T
     // Create a new exchange context.
     mpExchangeCtx = mpExchangeMgr->NewContext(session, this);
     VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
-    if (session.IsGroupSession())
-    {
-        // Exchange will be closed by WriteClientHandle::SendWriteRequest for group messages
-        err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::WriteRequest, std::move(packet),
-                                         Messaging::SendFlags(Messaging::SendMessageFlags::kNoAutoRequestAck));
-    }
-    else
-    {
-        mpExchangeCtx->SetResponseTimeout(timeout);
 
-        err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::WriteRequest, std::move(packet),
-                                         Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
-        SuccessOrExit(err);
+    mpExchangeCtx->SetResponseTimeout(timeout);
+
+    // kExpectResponse is ignored by ExchangeContext in case of groupcast
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::WriteRequest, std::move(packet),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    SuccessOrExit(err);
+
+    if (!session.IsGroupSession())
```

Comment:
Well yes, but this is handled when closing the `WriteClient` in `WriteClientHandle::SendWriteRequest` when calling `SetWriteClient(nullptr);`

Reply:
Thinking about this a bit more, I still think that in all cases of calling `SendWriteRequest`, that the ownership should always be transferred to the IM. This makes things a lot simpler.

Consequently, the actual task of closing out the EC should be on the WriteClient to do, as well as the subsequent Shutdown(). This ensures that the WriteClient (which knows expressly what its lifetime semantics should be) is responsible for that for all types of interaction variants.

Reply:
As in, move the logic out of the handle and into here?

Reply:
@jepenven-silabs Should probably have a comment here about how the handle will advance our state?

But better yet to do it all here and not branch on group bits in handle code.

Reply:
I could move the logic here instead of the handle if nothing prevent this. However as to transfer the ownership to the IM I do feel like this is a scope creep, because from what I saw, There is not much support in the IM to handle a Group context. @mrjerryjohns The goal is to complete a testable end-to-end group Attribute write as soon as possible and since there is stil a lot to do , I won't transfer the ownership to the IM in this PR. If you want, you are more than welcome to jump in the group communication feature and rectify this in  another PR 😄

Reply:
Done.